File: android-beans/src/test/java/rocks/frieler/android/beans/BeanDependencyFulfillmentTest.java
package rocks.frieler.android.beans;

import org.junit.Test;

import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static rocks.frieler.android.beans.BeanDependency.Fulfillment.FULFILLED;
import static rocks.frieler.android.beans.BeanDependency.Fulfillment.UNFULFILLED;
import static rocks.frieler.android.beans.BeanDependency.Fulfillment.UNFULFILLED_OPTIONAL;

public class BeanDependencyFulfillmentTest {
    @Test
    public void testUnfulfilledIsLowerThanUnfulfilledOptional() {
        assertThat(BeanDependency.Fulfillment.min(UNFULFILLED, UNFULFILLED_OPTIONAL), is(UNFULFILLED));
        assertThat(BeanDependency.Fulfillment.min(UNFULFILLED_OPTIONAL, UNFULFILLED), is(UNFULFILLED));
    }

    @Test
    public void testUnfulfilledIsLowerThanFulfilled() {
        assertThat(BeanDependency.Fulfillment.min(UNFULFILLED, FULFILLED), is(UNFULFILLED));
        assertThat(BeanDependency.Fulfillment.min(FULFILLED, UNFULFILLED), is(UNFULFILLED));
    }

    @Test
    public void testUnfulfilledOptionalIsLowerThanFulfilled() {
        assertThat(BeanDependency.Fulfillment.min(UNFULFILLED_OPTIONAL, FULFILLED), is(UNFULFILLED_OPTIONAL));
        assertThat(BeanDependency.Fulfillment.min(FULFILLED, UNFULFILLED_OPTIONAL), is(UNFULFILLED_OPTIONAL));
    }

    @Test
    public void testMinOfSameFulfillmentIsThatFulfillment() {
        assertThat(BeanDependency.Fulfillment.min(UNFULFILLED, UNFULFILLED), is(UNFULFILLED));
        assertThat(BeanDependency.Fulfillment.min(UNFULFILLED_OPTIONAL, UNFULFILLED_OPTIONAL), is(UNFULFILLED_OPTIONAL));
        assertThat(BeanDependency.Fulfillment.min(FULFILLED, FULFILLED), is(FULFILLED));
    }
}

File: android-beans/src/test/kotlin/rocks/frieler/android/beans/BeanDependencyFulfillmentTest.kt
package rocks.frieler.android.beans

import assertk.assertThat
import assertk.assertions.isEqualTo
import org.junit.Test
import rocks.frieler.android.beans.BeanDependency.Fulfillment.*
import rocks.frieler.android.beans.BeanDependency.Fulfillment.Comparison.min

class BeanDependencyFulfillmentTest {
	@Test
	fun `UNFULFILLED is lower than UNFULFILLED_OPTIONAL`() {
		assertThat(min(UNFULFILLED, UNFULFILLED_OPTIONAL)).isEqualTo(UNFULFILLED)
		assertThat(min(UNFULFILLED_OPTIONAL, UNFULFILLED)).isEqualTo(UNFULFILLED)
	}

	@Test
	fun `UNFULFILLED is lower than FULFILLED`() {
		assertThat(min(UNFULFILLED, FULFILLED)).isEqualTo(UNFULFILLED)
		assertThat(min(FULFILLED, UNFULFILLED)).isEqualTo(UNFULFILLED)
	}

	@Test
	fun `UNFULFILLED_OPTIONAL is lower than FULFILLED`() {
		assertThat(min(UNFULFILLED_OPTIONAL, FULFILLED)).isEqualTo(UNFULFILLED_OPTIONAL)
		assertThat(min(FULFILLED, UNFULFILLED_OPTIONAL)).isEqualTo(UNFULFILLED_OPTIONAL)
	}

	@Test
	fun `min() of same Fulfillment is that Fulfillment`() {
		assertThat(min(UNFULFILLED, UNFULFILLED)).isEqualTo(UNFULFILLED)
		assertThat(min(UNFULFILLED_OPTIONAL, UNFULFILLED_OPTIONAL)).isEqualTo(UNFULFILLED_OPTIONAL)
		assertThat(min(FULFILLED, FULFILLED)).isEqualTo(FULFILLED)
	}
}
